.github/workflows/lint-and-test.yaml

```diff
@@ -21,15 +23,78 @@ jobs:
             image: ubuntu-latest
 
     steps:
-      - uses: actions/checkout@v2
-      - run: pwd && docker-compose --version && docker --version
+      - uses: actions/checkout@v3
+
+      - name: Check version info
+        run: pwd && docker-compose --version && docker --version
+
+      - name: Set up Docker Buildx
+        if: ${{ env.USE_CACHE == 'true' }}
+        id: buildx
+        uses: docker/setup-buildx-action@v2
+        with:
+          driver-opts: network=host
+
+      - name: Cache Docker layers
+        if: ${{ env.USE_CACHE == 'true' }}
+        uses: actions/cache@v3
+        with:
+          path: /tmp/.buildx-cache/${{ matrix.python-version }}
+          key: buildx-${{ github.ref }}-${{ github.sha }}
+          restore-keys: |
+            buildx-${{ github.ref }}
+            buildx-
+
+      - name: Cache Docker registry
+        if: ${{ env.USE_CACHE == 'true' }}
+        uses: actions/cache@v3
+        with:
+          path: /tmp/docker-registry/${{ matrix.python-version }}
+          key: docker-registry-${{ github.ref }}-${{ github.sha }}
+          restore-keys: |
+            docker-registry-${{ github.ref }}
+            docker-registry-
+
+      - name: Boot-up local Docker registry
+        if: ${{ env.USE_CACHE == 'true' }}
+        run: docker run -d -p 5000:5000 --restart=always --name registry -v /tmp/docker-registry:/var/lib/registry registry:2
+      
+      - name: Wait for Docker registry
+        if: ${{ env.USE_CACHE == 'true' }}
+        run: npx wait-on tcp:5000
+
+      - name: Generate Docker image tag
+        run: |
+            SHA=${{ github.sha }}
+            TAG=$(TZ=UTC-9 date '+%Y%m')-${SHA:0:7}
+            echo "DOCKER_IMAGE_TAG_CI=$TAG" >> $GITHUB_ENV
+            echo TAG $TAG
+            echo "::set-output name=docker_image_tag_ci::$TAG"
+
+      - name: Build docker image with cache
+        if: ${{ env.USE_CACHE == 'true' }}
+        run: |
+          docker buildx bake \
+          --builder="${{ steps.buildx.outputs.name }}" \
+          --set="core.args.PYTHON_VERSION=${{ matrix.python-version }}" \
+          --set="*.cache-from=type=local,src=/tmp/.buildx-cache/${{ matrix.python-version }}" \
+          --set="*.cache-to=type=local,dest=/tmp/.buildx-cache-new/${{ matrix.python-version }}" \
+          --push \
+          -f docker-compose.yaml
+        working-directory: ${{ env.DOCKER_COMPOSE_DIRECTORY }}
 
       - name: Build docker image
+        if: ${{ env.USE_CACHE != 'true' }}
         run: docker-compose build --parallel --build-arg PYTHON_VERSION=${{ matrix.python-version }} core
         working-directory: ${{ env.DOCKER_COMPOSE_DIRECTORY }}
```

Comment:
Yes, you are right. But as I mentioned above, my intention is turn on/off cache by `USE_CACHE`, so `environments/ci` is used from both is intentional behavior! Because there is no cache related setting in `environments/ci`, it also works without cache. @YoshikiKubotani

Reply:
Thank you for your reply.
My understanding is that `ci/docker-compose.yaml` uses `localhost:5000/core:${DOCKER_IMAGE_TAG_CI}`, which is on the private repository.
I just thought this private repo is initiated at the `Boot-up local Docker registry` step, and this step would not run under `USE_CACHE = false` because it has `if ${{ env.USE_CACHE == 'true' }}`.
Does the docker on the VM run successfully even when the private repository is not working?

If I am saying something strange; sorry about that.  In that case, I appreciate it if you let me know.

Reply:
@YoshikiKubotani 
Thank you for your question. `localhost:5000/core:${DOCKER_IMAGE_TAG_CI}` represents image name and tag. When `docker push` is executed, name is parsed and target repo to push is decided. However, until `docker push` is executed, `localhost:5000/core` is just name of image, and `docker push` does not executed when cache is deactivated. So I think there is no problem.
I run CI job when `USE_CACHE = false` for confirmation.
 https://github.com/cvpaperchallenge/Ascender/actions/runs/3265274602/jobs/5367229940

Reply:
@gatheluck 
Thank you.
I misunderstood the `image` option in the docker-compose.
I thought that it specifies the actual building image, but it is actually just a name and tag.
Then, this pull request seems to have no problem.